File: host-windows/tests/AppControl.Host.Tests/CryptoVectorTests.cs
using System.Text;
using System.Text.Json;
using AppControl.Host.Security;
using NSec.Cryptography;
using Xunit;

namespace AppControl.Host.Tests;

/// <summary>
/// Prueft die C#-Kryptografie gegen die gemeinsamen Testvektoren aus
/// tools/crypto-vectors/vectors.json.
///
/// WARUM DAS WICHTIG IST: Dieselbe Krypto wird dreimal implementiert - in C#, in
/// Swift und in der Node-Referenz. Zwei unabhaengige Implementierungen einer
/// Spezifikation weichen erfahrungsgemaess in genau den Details voneinander ab,
/// die niemand testet: Byte-Reihenfolge des Zaehlers, Reihenfolge der Nonces im
/// Salt, Zuordnung der drei DH-Operationen zu den Rollen.
///
/// Ohne diese Tests aeussert sich so eine Abweichung als "Verbindung schlaegt
/// fehl, Fehlermeldung unbrauchbar" beim Nutzer. Mit ihnen als fehlschlagender
/// Test mit klarer Ursache.
/// </summary>
public sealed class CryptoVectorTests
{
    private static readonly JsonDocument Vectors = LoadVectors();

    private static JsonDocument LoadVectors()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "vectors.json");
        Assert.True(File.Exists(path),
            $"vectors.json fehlt unter {path}. " +
            "Erzeugen mit: node tools/crypto-vectors/generate.mjs > tools/crypto-vectors/vectors.json");
        return JsonDocument.Parse(File.ReadAllText(path));
    }

    private static byte[] Hex(string key, params string[] path)
    {
        var element = Vectors.RootElement;
        foreach (var segment in path) element = element.GetProperty(segment);
        return Convert.FromHexString(element.GetProperty(key).GetString()!);
    }

    [Fact]
    public void HandshakeKey_stimmt_mit_der_Referenz_ueberein()
    {
        var psk = Hex("psk", "pairing");
        var roomId = Hex("roomId", "pairing");
        var expected = Hex("handshakeKey", "pairing");

        var actual = Handshake.DeriveHandshakeKey(psk, roomId);

        Assert.Equal(expected, actual);
    }

    [Fact]
    public void SessionKey_Host_stimmt_mit_der_Referenz_ueberein()
    {
        var result = ComputeAs(HandshakeRole.Host);
        Assert.Equal(Hex("sessionKey", "handshake"), result.SessionKey);
    }

    [Fact]
    public void SessionKey_Viewer_ergibt_denselben_Schluessel_wie_Host()
    {
        // Die zentrale Eigenschaft des Handshakes: Zwei Seiten mit
        // UNTERSCHIEDLICHEN privaten Schluesseln kommen auf DENSELBEN
        // Sitzungsschluessel.
        var host = ComputeAs(HandshakeRole.Host);
        var viewer = ComputeAs(HandshakeRole.Viewer);

        Assert.Equal(host.SessionKey, viewer.SessionKey);
    }

    private static HandshakeResult ComputeAs(HandshakeRole role)
    {
        var x = Vectors.RootElement.GetProperty("x25519");
        byte[] Get(string name) => Convert.FromHexString(x.GetProperty(name).GetString()!);

        var (ephPriv, staticPriv, peerEph, peerStatic) = role == HandshakeRole.Host
            ? (Get("hostEphPriv"), Get("hostStaticPriv"), Get("viewerEphPub"), Get("viewerStaticPub"))
            : (Get("viewerEphPriv"), Get("viewerStaticPriv"), Get("hostEphPub"), Get("hostStaticPub"));

        var import = (byte[] raw) => Key.Import(
            KeyAgreementAlgorithm.X25519, raw, KeyBlobFormat.RawPrivateKey,
            new KeyCreationParameters { ExportPolicy = KeyExportPolicies.AllowPlaintextExport });

        return Handshake.ComputeSession(
            role, import(ephPriv), import(staticPriv), peerEph, peerStatic,
            nonceHost: Hex("nonceHost", "handshake"),
            nonceViewer: Hex("nonceViewer", "handshake"),
            peerName: "test");
    }

    [Fact]
    public void SAS_stimmt_in_Indizes_und_Symbolen_mit_der_Referenz_ueberein()
    {
        var sessionKey = Hex("sessionKey", "handshake");
        var expectedIndices = Vectors.RootElement.GetProperty("sas").GetProperty("indices")
            .EnumerateArray().Select(e => e.GetInt32()).ToArray();
        var expectedEmoji = Vectors.RootElement.GetProperty("sas").GetProperty("emoji")
            .EnumerateArray().Select(e => e.GetString()!).ToArray();

        var (emoji, indices) = Handshake.ComputeSas(sessionKey);

        Assert.Equal(expectedIndices, indices);
        Assert.Equal(string.Join(' ', expectedEmoji), emoji);
    }

    [Fact]
    public void EmojiTabelle_hat_64_eindeutige_Symbole()
    {
        // Die Reihenfolge ist Teil des Protokolls - eine Aenderung wuerde dazu
        // fuehren, dass Host und Viewer verschiedene Symbole fuer denselben
        // Schluessel zeigen und die Nutzer eine gesunde Verbindung abbrechen.
        Assert.Equal(64, EmojiTable.Symbols.Length);
        Assert.Equal(64, EmojiTable.Symbols.Distinct().Count());
    }

    [Fact]
    public void Umschlag_reproduziert_den_Referenz_Ciphertext_bitgenau()
    {
        var key = Hex("sessionKey", "handshake");
        var env = Vectors.RootElement.GetProperty("envelope");
        var salt4 = Convert.FromHexString(env.GetProperty("salt4").GetString()!);
        var plaintext = Convert.FromBase64String(env.GetProperty("plaintext").GetString()!);
        var expected = Convert.FromBase64String(env.GetProperty("sealed").GetString()!);

        var channel = new SecureChannel(key, ChannelPhase.Session, ChannelDirection.HostToViewer);
        var actual = channel.SealWithFixedNonce(plaintext, counter: 1, salt4);

        Assert.Equal(expected, actual);
    }

    [Fact]
    public void Umschlag_der_Gegenrichtung_laesst_sich_oeffnen()
    {
        var key = Hex("sessionKey", "handshake");
        var expected = Convert.FromBase64String(
            Vectors.RootElement.GetProperty("envelope").GetProperty("plaintext").GetString()!);
        var sealedBytes = Convert.FromBase64String(
            Vectors.RootElement.GetProperty("envelope").GetProperty("sealed").GetString()!);

        // Der Referenzumschlag ist host->viewer; ihn oeffnet ein Kanal, der
        // selbst viewer->host sendet.
        var viewerSide = new SecureChannel(key, ChannelPhase.Session, ChannelDirection.ViewerToHost);

        Assert.Equal(expected, viewerSide.Open(sealedBytes));
    }

    [Fact]
    public void Manipulierter_Ciphertext_wird_abgewiesen()
    {
        var key = Hex("sessionKey", "handshake");
        var tampered = Convert.FromBase64String(
            Vectors.RootElement.GetProperty("envelope").GetProperty("sealed").GetString()!);
        tampered[25] ^= 0x01;

        var viewerSide = new SecureChannel(key, ChannelPhase.Session, ChannelDirection.ViewerToHost);

        Assert.Null(viewerSide.Open(tampered));
    }

    [Fact]
    public void Gefaelschte_Richtung_im_Header_wird_abgewiesen()
    {
        // Der Header ist Associated Data der AEAD. Eine geaenderte Richtung
        // bricht die Authentifizierung - das verhindert Reflection-Angriffe,
        // bei denen eine an uns gesendete Nachricht zurueckgespiegelt wird.
        var key = Hex("sessionKey", "handshake");
        var tampered = Convert.FromBase64String(
            Vectors.RootElement.GetProperty("envelope").GetProperty("sealed").GetString()!);
        tampered[5] = (byte)ChannelDirection.ViewerToHost;

        var viewerSide = new SecureChannel(key, ChannelPhase.Session, ChannelDirection.ViewerToHost);

        Assert.Null(viewerSide.Open(tampered));
    }

    [Fact]
    public void Replay_desselben_Umschlags_wird_abgewiesen()
    {
        var key = Hex("sessionKey", "handshake");
        var sealedBytes = Convert.FromBase64String(
            Vectors.RootElement.GetProperty("envelope").GetProperty("sealed").GetString()!);

        var viewerSide = new SecureChannel(key, ChannelPhase.Session, ChannelDirection.ViewerToHost);

        Assert.NotNull(viewerSide.Open(sealedBytes));
        Assert.Null(viewerSide.Open(sealedBytes));   // zweites Mal: Replay
    }

    [Fact]
    public void Falscher_Schluessel_oeffnet_den_Umschlag_nicht()
    {
        var wrong = Hex("sessionKey", "handshake");
        wrong[31] ^= 0xFF;
        var sealedBytes = Convert.FromBase64String(
            Vectors.RootElement.GetProperty("envelope").GetProperty("sealed").GetString()!);

        var channel = new SecureChannel(wrong, ChannelPhase.Session, ChannelDirection.ViewerToHost);

        Assert.Null(channel.Open(sealedBytes));
    }
}

File: viewer-macos/Tests/AppControlViewerTests/CryptoVectorTests.swift
import Crypto
import XCTest
@testable import AppControlViewer

/// Prüft die Swift-Kryptografie gegen die gemeinsamen Testvektoren aus
/// `tools/crypto-vectors/vectors.json`.
///
/// **Warum das wichtig ist:** Dieselbe Krypto wird dreimal implementiert — in C#,
/// in Swift und in der Node-Referenz. Zwei unabhängige Implementierungen einer
/// Spezifikation weichen erfahrungsgemäß in genau den Details voneinander ab, die
/// niemand testet: Byte-Reihenfolge des Zählers, Reihenfolge der Nonces im Salt,
/// Zuordnung der drei DH-Operationen zu den Rollen.
///
/// Ohne diese Tests äußert sich so eine Abweichung als „Verbindung schlägt fehl,
/// Fehlermeldung unbrauchbar" beim Nutzer. Mit ihnen als fehlschlagender Test mit
/// klarer Ursache.
final class CryptoVectorTests: XCTestCase {

    private struct Vectors: Decodable {
        struct X25519: Decodable {
            let hostStaticPriv, hostStaticPub, hostEphPriv, hostEphPub: String
            let viewerStaticPriv, viewerStaticPub, viewerEphPriv, viewerEphPub: String
        }
        struct Pairing: Decodable { let psk, roomId, handshakeKey: String }
        struct HandshakeVec: Decodable {
            let nonceHost, nonceViewer, dh1, dh2, dh3, salt, sessionKey: String
        }
        struct Sas: Decodable { let bytes: String; let indices: [Int]; let emoji: [String] }
        struct Envelope: Decodable {
            let phase, direction: Int
            let counter: Int
            let salt4, header, plaintext, sealed: String
        }

        let x25519: X25519
        let pairing: Pairing
        let handshake: HandshakeVec
        let sas: Sas
        let envelope: Envelope
    }

    private static var vectors: Vectors = {
        guard let url = Bundle.module.url(forResource: "vectors", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let decoded = try? JSONDecoder().decode(Vectors.self, from: data) else {
            fatalError("""
                vectors.json konnte nicht geladen werden.
                Erzeugen mit: node tools/crypto-vectors/generate.mjs > tools/crypto-vectors/vectors.json
                und nach viewer-macos/Tests/AppControlViewerTests/ kopieren oder verlinken.
                """)
        }
        return decoded
    }()

    private var v: Vectors { Self.vectors }

    private func hex(_ string: String) -> Data {
        var data = Data(capacity: string.count / 2)
        var index = string.startIndex
        while index < string.endIndex {
            let next = string.index(index, offsetBy: 2)
            data.append(UInt8(string[index..<next], radix: 16)!)
            index = next
        }
        return data
    }

    private func hexString(_ data: Data) -> String {
        data.map { String(format: "%02x", $0) }.joined()
    }

    // ── Handshake ────────────────────────────────────────────────────────────

    func testHandshakeKeyMatchesReference() throws {
        let derived = Handshake.deriveHandshakeKey(
            psk: hex(v.pairing.psk),
            roomId: hex(v.pairing.roomId))

        XCTAssertEqual(hexString(derived), v.pairing.handshakeKey)
    }

    func testSessionKeyAsViewerMatchesReference() throws {
        let result = try computeSession(role: .viewer)
        XCTAssertEqual(hexString(result.sessionKey), v.handshake.sessionKey)
    }

    func testSessionKeyAsHostMatchesReference() throws {
        // Der Viewer berechnet in der Praxis nur die Viewer-Seite, aber beide
        // Pfade müssen stimmen — sonst würde ein Rollentausch (etwa für Tests
        // oder eine spätere bidirektionale Variante) still falsche Ergebnisse
        // liefern.
        let result = try computeSession(role: .host)
        XCTAssertEqual(hexString(result.sessionKey), v.handshake.sessionKey)
    }

    func testHostAndViewerDeriveIdenticalKey() throws {
        // Die zentrale Eigenschaft des Handshakes: Zwei Seiten mit
        // UNTERSCHIEDLICHEN privaten Schlüsseln kommen auf DENSELBEN
        // Sitzungsschlüssel.
        let host = try computeSession(role: .host)
        let viewer = try computeSession(role: .viewer)

        XCTAssertEqual(host.sessionKey, viewer.sessionKey)
    }

    private func computeSession(role: HandshakeRole) throws -> HandshakeResult {
        let x = v.x25519

        let (ephPriv, staticPriv, peerEph, peerStatic): (String, String, String, String) =
            role == .host
                ? (x.hostEphPriv, x.hostStaticPriv, x.viewerEphPub, x.viewerStaticPub)
                : (x.viewerEphPriv, x.viewerStaticPriv, x.hostEphPub, x.hostStaticPub)

        return try Handshake.computeSession(
            role: role,
            ownEphemeral: Curve25519.KeyAgreement.PrivateKey(rawRepresentation: hex(ephPriv)),
            ownStatic: Curve25519.KeyAgreement.PrivateKey(rawRepresentation: hex(staticPriv)),
            peerEphemeralPublic: hex(peerEph),
            peerStaticPublic: hex(peerStatic),
            nonceHost: hex(v.handshake.nonceHost),
            nonceViewer: hex(v.handshake.nonceViewer),
            peerName: "test")
    }

    // ── SAS ──────────────────────────────────────────────────────────────────

    func testSasMatchesReference() {
        let (emoji, indices) = Handshake.computeSas(sessionKey: hex(v.handshake.sessionKey))

        XCTAssertEqual(indices, v.sas.indices)
        XCTAssertEqual(emoji, v.sas.emoji.joined(separator: " "))
    }

    func testEmojiTableHas64UniqueSymbols() {
        // Die Reihenfolge ist Teil des Protokolls — eine Änderung würde dazu
        // führen, dass Host und Viewer verschiedene Symbole für denselben
        // Schlüssel zeigen und die Nutzer eine gesunde Verbindung abbrechen.
        XCTAssertEqual(EmojiTable.symbols.count, 64)
        XCTAssertEqual(Set(EmojiTable.symbols).count, 64)
    }

    // ── Umschlag ─────────────────────────────────────────────────────────────

    func testSealReproducesReferenceCiphertextExactly() throws {
        let key = SymmetricKey(data: hex(v.handshake.sessionKey))

        let sealed = try SecureChannel.seal(
            key: key, phase: .session, direction: .hostToViewer,
            counter: UInt64(v.envelope.counter),
            salt4: hex(v.envelope.salt4),
            plaintext: Data(base64Encoded: v.envelope.plaintext)!)

        XCTAssertEqual(sealed.base64EncodedString(), v.envelope.sealed)
    }

    func testOpenAcceptsReferenceEnvelope() {
        // Der Referenzumschlag ist host→viewer; ihn öffnet ein Kanal, der selbst
        // viewer→host sendet.
        let channel = SecureChannel(
            key: hex(v.handshake.sessionKey), phase: .session, sendDirection: .viewerToHost)

        let opened = channel.open(Data(base64Encoded: v.envelope.sealed)!)

        XCTAssertEqual(opened?.base64EncodedString(), v.envelope.plaintext)
    }

    func testTamperedCiphertextIsRejected() {
        var tampered = Data(base64Encoded: v.envelope.sealed)!
        tampered[25] ^= 0x01

        let channel = SecureChannel(
            key: hex(v.handshake.sessionKey), phase: .session, sendDirection: .viewerToHost)

        XCTAssertNil(channel.open(tampered))
    }

    func testTamperedHeaderIsRejected() {
        // Der Header ist Associated Data der AEAD. Eine geänderte Richtung bricht
        // die Authentifizierung — das verhindert Reflection-Angriffe, bei denen
        // eine an uns gesendete Nachricht zurückgespiegelt wird.
        var tampered = Data(base64Encoded: v.envelope.sealed)!
        tampered[5] = ChannelDirection.viewerToHost.rawValue

        let channel = SecureChannel(
            key: hex(v.handshake.sessionKey), phase: .session, sendDirection: .viewerToHost)

        XCTAssertNil(channel.open(tampered))
    }

    func testReplayIsRejected() {
        let channel = SecureChannel(
            key: hex(v.handshake.sessionKey), phase: .session, sendDirection: .viewerToHost)
        let envelope = Data(base64Encoded: v.envelope.sealed)!

        XCTAssertNotNil(channel.open(envelope))
        XCTAssertNil(channel.open(envelope), "Ein zweites Mal ist ein Replay")
    }

    func testWrongKeyCannotOpen() {
        var wrong = hex(v.handshake.sessionKey)
        wrong[31] ^= 0xFF

        let channel = SecureChannel(key: wrong, phase: .session, sendDirection: .viewerToHost)

        XCTAssertNil(channel.open(Data(base64Encoded: v.envelope.sealed)!))
    }
}
